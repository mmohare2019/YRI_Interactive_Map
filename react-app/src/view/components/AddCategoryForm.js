// pop-up form for adding categories
import React from "react"
import {Navigate} from "react-router-dom"
import Form from 'react-bootstrap/Form'; 
import Button from 'react-bootstrap/Button';
import AdminHeader from "./AdminHeader"
const categoryHandler = require("../../event-handler/categoryHandler")

export default class AddCategoryForm extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            categoryName: "",
            color: "#000000",
            icon: "",
            iconFile: "",
            errMsg: "",
            backToCategoryScreen: false
        }

        this.handleChange = this.handleChange.bind(this)
        this.handleSubmit = this.handleSubmit.bind(this)
    }

    handleChange(event) {
        const target = event.target
        const value = target.value
        const name = target.name

        this.setState({
            [name]: value,
        })
    }

    async handleSubmit(e) {
        e.preventDefault()

        console.log(this.state.categoryName)
        console.log(this.state.color)
        console.log(this.state.iconFile.name)

        var errMsg = ""
        var backToCategoryScreen= false

        if(this.state.name === "") {
            errMsg = "Name cannot be empty"
        } else {
            try {
                var response = await categoryHandler.submitAddCategoryForm(
                    this.state.categoryName, this.state.color, this.state.iconFile)
            
                if(response.status === 201) {
                    backToCategoryScreen = true
                } else {
                    errMsg = "Category name already exists"
                }
            }
            catch(error) {
                errMsg = "Category name already exists"
            }
        }

        this.setState({
            errMsg: errMsg,
            backToCategoryScreen: backToCategoryScreen
        })
    }

    render() {
        let backToCategoryScreen = this.state.backToCategoryScreen

        return(<>
            {backToCategoryScreen && (
                <Navigate to="/category" replace={true} />
            )}

            <AdminHeader/>

            <h1 className="title"> Add Category </h1> 

            <Form onSubmit={this.handleSubmit} className="container">
                <Form.Group className="mb-3" controlId="name">
                    <Form.Label>Name</Form.Label>
                    <Form.Control data-testid="name" type="text" name="categoryName" value={this.state.categoryName} placeholder="Enter name" onChange={this.handleChange}/>
                </Form.Group>
    
                <Form.Group className="mb-3" controlId="color">                       
                    <Form.Label>Color</Form.Label>
                    <Form.Control data-testid="color" type="color" name="color" value={this.state.color} onChange={this.handleChange}/>
                </Form.Group>
    
                <Form.Group className="mb-3" controlId="description">
                    <Form.Label>Icon</Form.Label>
                    <Form.Control 
                        type="file" 
                        accept="image/*" 
                        filename={this.state.icon} 
                        onChange={e => this.setState({"iconFile": e.target.files[0]})}/>
                </Form.Group>
                    
                <div className="text-center">
                    <Button variant="primary" type="submit" data-testid="submit">
                        Submit
                    </Button>
                </div>


                {this.state.errMsg !== "" && 
                    <label className="form-error-msg">
                        {this.state.errMsg}
                    </label>
                }
            </Form>
        </>)
    }
}