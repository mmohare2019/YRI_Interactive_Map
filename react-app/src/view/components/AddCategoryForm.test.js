import TestRenderer from "react-test-renderer";
import AddCategoryForm from "./AddCategoryForm";
import { fireEvent, screen, render} from "@testing-library/react"
import React from "react";
import renderer from 'react-test-renderer';

const categoryHandler = require("../../event-handler/categoryHandler")


describe("add category form tests", () => {
    test("render smoke test", () => {
        TestRenderer.create(<AddCategoryForm/>)
    })
});

test("add category form test", function () {
    const component = renderer.create(<AddCategoryForm/>).toJSON();
    expect(component).toMatchSnapshot(); 

    render(<AddCategoryForm/>)
    const nameInput = screen.getByPlaceholderText("Enter name");
    fireEvent.change(nameInput, {target: {value: "test name"}});

    const colorInput = screen.getByTestId("color");
    fireEvent.change(colorInput, {target: {value: "#692f89"}})

    const submitInput = screen.getByTestId("submit")

    categoryHandler.submitAddCategoryForm = async(name, color, icon) => {
        return {
            status: 200
        }
    } 
        
    fireEvent.click(submitInput)
});
