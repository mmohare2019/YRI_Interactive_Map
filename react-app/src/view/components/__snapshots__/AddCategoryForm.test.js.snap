// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add category form test 1`] = `
Array [
  <nav
    className="header navbar navbar-expand navbar-light"
  >
    <div
      className="container"
    >
      <a
        className="navbar-brand"
        href="/admin"
      >
        <img
          alt="YRI logo"
          className="yri-logo"
          height="30"
          src="YRI_LOGO.png"
          width="30"
        />
        <img
          alt="CCSJ logo"
          className="ccsj-logo"
          height="30"
          src="ccsjLogo.jpg"
          width="30"
        />
        <button
          aria-controls="navbarNavAltMarkup"
          aria-expanded="false"
          aria-label="Toggle navigation"
          class="navbar-toggler"
          data-target="#navbarNavAltMarkup"
          data-toggle="collapse"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="collapse navbar-collapse"
          id="navbarNavAltMarkup"
        >
          <div
            class="navbar-nav"
          >
            <a
              class="nav-item nav-link"
              href="/admin"
            >
              Home
            </a>
            <a
              class="nav-item nav-link"
              href="/inbox"
            >
              Messages
            </a>
            <a
              class="nav-item nav-link"
              href="/partners"
            >
              Partners
            </a>
            <a
              class="nav-item nav-link"
              href="/category"
            >
              Categories
            </a>
          </div>
        </div>
      </a>
    </div>
  </nav>,
  <br />,
  <h1
    className="title"
  >
     Add Category 
  </h1>,
  <form
    className="container"
    onSubmit={[Function]}
  >
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="name"
      >
        Name
      </label>
      <input
        className="form-control"
        data-testid="name"
        id="name"
        name="categoryName"
        onChange={[Function]}
        placeholder="Enter name"
        type="text"
        value=""
      />
    </div>
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="color"
      >
        Color
      </label>
      <input
        className="form-control form-control-color"
        data-testid="color"
        id="color"
        name="color"
        onChange={[Function]}
        type="color"
        value="#000000"
      />
    </div>
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="description"
      >
        Icon
      </label>
      <input
        accept="image/*"
        className="form-control"
        filename=""
        id="description"
        onChange={[Function]}
        type="file"
      />
    </div>
    <div
      className="text-center"
    >
      <button
        className="btn btn-primary"
        data-testid="submit"
        disabled={false}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>,
]
`;
